### PR TITLE
feat: add read_thumbnail

### PR DESCRIPTION
**Pull request by fabien-brulport**
Add a thumbnail method similar to openslide's one

### CODE REVIEW

File: src/view.rs [hunk 1]
```diff
@@ -125,4 +123,62 @@ impl<'a> View<'a> {
...
+impl Size {
+    pub fn new(w: u32, h: u32) -> Self {
+        Self { w, h }
+    }
+    pub fn from_dimensions_range(range: &DimensionsRange) -> Self {
+        Self {
+            w: (range.end_x - range.start_x) / range.step_x,
+            h: (range.end_y - range.start_y) / range.step_y,
+        }
+    }
```
Comment (fabien-brulport) [inline src/view.rs:183]: could be moved elsewhere ?

File: src/utils.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
+use crate::{Result, Size};
+use std::cmp;
+
+#[cfg(feature = "image")]
+use fast_image_resize as fr;
+#[cfg(feature = "image")]
+use {crate::errors::ImageError, image::RgbImage};
+
+#[cfg(feature = "image")]
+pub fn resize_rgb_image(image: RgbImage, new_size: &Size) -> Result<RgbImage> {
```
Comment (AzHicham) [inline src/utils.rs:10]: pub(crate) instead of pub ?

File: src/utils.rs [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    resizer
+        .resize(&src_image, &mut dst_image, &option)
+        .map_err(|err| ImageError::Other(err.to_string()))?;
+    let image = RgbImage::from_vec(new_size.w, new_size.h, dst_image.into_vec()).unwrap(); // safe because dst_image buffer is big enough
+
+    Ok(image)
+}
+
+#[cfg(feature = "image")]
+pub fn preserve_aspect_ratio(size: &Size, dimension: &Size) -> Size {
```
Comment (AzHicham) [inline src/utils.rs:35]: pub fn -> fn ?

File: src/view.rs [hunk 4]
```diff
@@ -1,14 +1,12 @@
 //! This module contains all functions related to Philips Views
 //!
 
+use crate::utils::{preserve_aspect_ratio, resize_rgb_image};
```
Comment (AzHicham) [inline src/view.rs:4]: #[cfg(feature = "image")]

File: src/view.rs [hunk 5]
```diff
@@ -125,4 +125,119 @@ impl<'a> View<'a> {
...
+        let image = self.read_image(engine, &region_request)?;
+        let final_size =
+            preserve_aspect_ratio(&size, &Size::from_dimensions_range(&dimensions_range));
+        let image = resize_rgb_image(image, &final_size)?;
+        Ok(image)
+    }
+
+    // Get the appropriate level for the given dimensions: i.e. the level with at least one
+    // dimensions greater than the dimension requested along one axis
+    pub fn get_best_level_for_dimensions(&self, dimension: &Size) -> Result<u32> {
```
[the remark this answers is not in the document]
Comment (fabien-brulport) [reply src/view.rs:152]: ok but can I import the fixture here (in the mod test) ? As I will need it to have a View

[the remark this answers is not in the document]
Comment (AzHicham) [reply src/view.rs:152]: Yeah :) you can import in the mod tests

[the remark this answers is not in the document]
Comment (fabien-brulport) [reply src/view.rs]: in the end I moved the code  to a more unitary function in utils, seems better IMO 🙂

File: src/view.rs (hunk 1 above)
Comment (fabien-brulport) [reply src/view.rs]: done !

File: src/utils.rs (hunk 3 above)
Comment (fabien-brulport) [reply src/utils.rs]: done !

File: src/utils.rs (hunk 2 above)
Comment (fabien-brulport) [reply src/utils.rs]: done !

File: src/view.rs (hunk 4 above)
Comment (fabien-brulport) [reply src/view.rs]: done !

File: src/view.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (AzHicham) [reply src/view.rs]: Perfect

File: src/bindings.rs [hunk 6]
```diff
@@ -125,6 +125,18 @@ pub(crate) mod ffi {
     }
 }
 
+impl ffi::Size {
+    pub fn new(w: u32, h: u32) -> Self {
+        Self { w, h }
+    }
+    pub fn from_dimensions_range(range: &ffi::DimensionsRange) -> Self {
```
Comment (AzHicham) [inline src/bindings.rs:132]: Sorry to bother you with this but this function may fail if end < start.
You can impl TryFrom<DimensionRange> for Size

File: src/utils.rs [hunk 7]
```diff
@@ -0,0 +1,163 @@
+use crate::{Result, Size};
+use std::cmp;
+
+#[cfg(feature = "image")]
+use fast_image_resize as fr;
+#[cfg(feature = "image")]
```
Comment (AzHicham) [inline src/utils.rs:6]: IMO ou can remove all cfg feature image here and put it on mod utils;

File: src/utils.rs [hunk 8]
```diff
@@ -0,0 +1,163 @@
...
+pub fn get_best_level_for_dimensions(
+    dimension: &Size,
+    dimension_level_0: &Size,
+    level_count: u32,
+) -> u32 {
+    let downsample = f64::max(
+        f64::from(dimension_level_0.w) / f64::from(dimension.w),
+        f64::from(dimension_level_0.h) / f64::from(dimension.h),
+    );
+    let level_dowsamples: Vec<f64> = (0..level_count)
```
Comment (AzHicham) [inline src/utils.rs:21]: Here is a more Rustish way to achieve that https://github.com/AzHicham/bioformats-rs/blob/714882e1b14a9fb2f4d1c2b5a6a3a6c4f59e7e3d/src/extra/wsi/slide.rs#L98
(0..).map().take_while().enumerate().rfind().unwrap_or() 
I’m not sure but it should works ^^

Comment (AzHicham) [reply src/utils.rs:21]: In biofirmats i think the code is broken we need to check against the first value (max slide dimension ie downsample of 1) then unwraps with the last value :/

**Comment by AzHicham** [comment]
For the benchmark I’ll merge the PR and fix the workflow later :)

File: src/utils.rs (hunk 7 above)
Comment (fabien-brulport) [reply src/utils.rs:6]: ok thanks ! yes I wanted to asking you if I could put the cfg feature on an entire mod 🙂

File: src/utils.rs (hunk 8 above)
Comment (fabien-brulport) [reply src/utils.rs:21]: the implementation looks correct to me !

File: src/bindings.rs (hunk 6 above)
Comment (fabien-brulport) [reply src/bindings.rs]: updated, If you have suggestions on how to better handle the Err do not hesitate 🙂

File: src/utils.rs (hunk 8 above)
Comment (fabien-brulport) [reply src/utils.rs]: I used the one from bioformats

File: src/bindings.rs [hunk 9]
```diff
@@ -125,6 +127,39 @@ pub(crate) mod ffi {
     }
 }
 
+impl ffi::Size {
+    pub fn new(w: u32, h: u32) -> Self {
+        Self { w, h }
+    }
+}
+impl TryFrom<&ffi::DimensionsRange> for ffi::Size {
+    type Error = PhilipsSlideError;
```
Comment (AzHicham) [inline src/bindings.rs:136]: here you can create a new type error 
#[derive(Error)]
DimensionsRangeToSizeError{
   StepXNull,
   StepXNull,
   EndXGreaterthanStartX,
   EndYGreaterthanStartY,
}

Then add this error into PhilipsSlideError 
#[transparent]
DimensionsRangeToSize(#[from] DimensionsRangeToSizeError)

Naming is bad but you get the idea haha

**Review by AzHicham — APPROVED** [review]
Nice GG !!

**Pull request merged**